File: array/sliding_window/fixed_size/0643_max_avg_subarray_i.js
/*
You are given an integer array nums consisting of n elements, and an integer k.

Find a contiguous subarray whose length is equal to k that has the maximum 
average value and return this value. Any answer with a calculation error 
less than 10-5 will be accepted.
*/

// sliding window & two pointers
// time complexity O(n)
// space complexity O(1)
var findMaxAverage = function (nums, k) {
  if (nums.length <= 1) return nums;
  let left = 0;
  let right = 0;
  let windowSum = 0;
  let avgMax = -Infinity;
  while (right < nums.length) {
    windowSum += nums[right];
    if (right - left + 1 >= k) {
      avgMax = Math.max(avgMax, windowSum / k);
      windowSum -= nums[left];
      left++;
    }
    right++;
  }
  return avgMax;
};

File: array/sliding_window/fixed_size/0643_max_avg_subarray_i.py
# You are given an integer array nums consisting of n elements, and an integer k.

# Find a contiguous subarray whose length is equal to k that has the maximum average value and return this value. Any answer with a calculation error less than 10-5 will be accepted.

# https://leetcode.com/problems/maximum-average-subarray-i/

class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        left = right = window_sum = 0
        l, ans = len(nums), float('-inf')
        while right < l:
            window_sum += nums[right]
            if right - left + 1 == k:
                ans = max(ans, window_sum)
                window_sum -= nums[left]
                left += 1
            right += 1
        return ans / k